### PR TITLE
fix(dynamics): FusedStage single-stage convergence causes duplicate writes and sink overflow

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit Pull Request

## Description

When a `FusedStage` has a single sub-stage (or when the last sub-stage declares a `convergence_hook`), no migration hook was auto-registered to graduate converged samples to `exit_status`. This caused `ConvergedSnapshotHook` to write the same converged samples on every subsequent step, eventually overflowing bounded sinks like `HostMemory`. Additionally, `ON_CONVERGE` hooks could fire for samples that were not active in the corresponding sub-stage.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes https://github.com/NVIDIA/nvalchemi-toolkit/issues/62

## Changes Made

- Extended `FusedStage.__init__` auto-registration loop from `range(len(sub_stages) - 1)` to `range(len(sub_stages))` so the last sub-stage gets a migration hook to `exit_status` when it declares a `convergence_hook`
- Captured `stage_active_masks` during the pre-update phase and intersected them with convergence results in `_step_impl`, ensuring `ON_CONVERGE` hooks only fire for samples that belong to the corresponding sub-stage
- Added four new tests: single-stage auto-registration, skip when no convergence hook, end-to-end snapshot no-overflow, and active-sample-only ON_CONVERGE mask verification

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

All 84 tests in `test/dynamics/test_single_loop.py` pass (including 4 new), plus 132 tests in `test_inflight.py`, `test_observer_hooks.py`, and `test_demo_dynamics.py` with zero regressions.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

The reproduction script `repro_fusedstage_duplicate_converged_writes.py` (untracked) confirms the fix: the "bad" scenario no longer overflows.

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemi/dynamics/base.py [hunk 1]
```diff
@@ -2916,10 +2925,26 @@ def _step_impl(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
...
+            stage_converged[converged] = True
+            stage_converged &= active_mask
+
+            if not stage_converged.any():
+                dynamics._last_converged = None
+                continue
+
+            dynamics._last_converged = torch.where(stage_converged)[0]
+            if dynamics._last_converged is not None:
                 dynamics._call_hooks(DynamicsStage.ON_CONVERGE, batch)
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemi/dynamics/base.py]: Fixed in bf1cea8791707fe70a23cb8c812c931f03b591cf

**Review by laserkelvin — APPROVED** [review]
LGTM
